### PR TITLE
[bug 443574] Added interface IFileSystemAccessExtension4

### PR DESCRIPTION
**Pull request by kthoms**
This patch introduces interface IFileSystemAccessExtension4, which defines method exists().
Implemented new exists() method for FSA implementations.
Adding test code.

Change-Id: I4380ebdabe023d5abeb0219fc8f6b27c73e01d4d
Signed-off-by: Karsten Thoms <karsten.thoms@itemis.de>

### CODE REVIEW

**Comment by oehme** [comment]
Jenkins test

**Comment by szarnekow** [comment]
Could you please update the @since tags to 2.9?

**Comment by kthoms** [comment]
Updated @since tags

File: plugins/org.eclipse.xtext/src/org/eclipse/xtext/generator/JavaIoFileSystemAccess.java [hunk 1]
```diff
@@ -235,4 +235,10 @@ public CharSequence readTextFile(String fileName, String outputCfgName) throws R
 			throw new RuntimeIOException(e);
 		}
 	}
+
+	@Override
+	public boolean exists(String fileName, String outputConfigurationName) throws RuntimeIOException {
+		File file = getFile(fileName, outputConfigurationName);
+		return file!=null && file.exists();
```
Comment (szarnekow) [inline plugins/org.eclipse.xtext/src/org/eclipse/xtext/generator/JavaIoFileSystemAccess.java]: This logic seems to be different from the logic based on InputStreams, e.g. file.exists returns true for directories.

Comment (kthoms) [inline plugins/org.eclipse.xtext/src/org/eclipse/xtext/generator/JavaIoFileSystemAccess.java]: In other words,
    && !file.isDirectory()
should be appended to the condition, right?

Comment (szarnekow) [inline plugins/org.eclipse.xtext/src/org/eclipse/xtext/generator/JavaIoFileSystemAccess.java]: Yes, that should be correct and semantically equivalent to IFile.exists. Could you add a test for that case, too and add a remark to the JavaDoc?

Comment (kthoms) [inline plugins/org.eclipse.xtext/src/org/eclipse/xtext/generator/JavaIoFileSystemAccess.java]: Will do when I have some time again.

File: intellij/org.eclipse.xtext.core.idea.tests/xtend-gen/org/eclipse/xtext/index/idea/completion/IndexTestLanguageCompletionContributor.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package org.eclipse.xtext.index.idea.completion;
```
Comment (szarnekow) [inline intellij/org.eclipse.xtext.core.idea.tests/xtend-gen/org/eclipse/xtext/index/idea/completion/IndexTestLanguageCompletionContributor.java]: This file should not be part of the PR

File: intellij/org.eclipse.xtext.idea/xtend-gen/org/eclipse/xtext/idea/extensions/IdeaProjectExtensions.java [hunk 3]
```diff
@@ -1,14 +1,12 @@
 package org.eclipse.xtext.idea.extensions;
```
Comment (szarnekow) [inline intellij/org.eclipse.xtext.idea/xtend-gen/org/eclipse/xtext/idea/extensions/IdeaProjectExtensions.java]: This file should not be part of the PR

File: plugins/org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/ParallelFileSystemAccess.xtend [hunk 4]
```diff
@@ -10,17 +10,18 @@ import org.eclipse.xtext.generator.IFileSystemAccessExtension2
...
 
 /**
  * @author Anton Kosyakov
  * @since 2.7
  */
-class ParallelFileSystemAccess implements IFileSystemAccess, IFileSystemAccessExtension, IFileSystemAccessExtension2, IFileSystemAccessExtension3 {
+class ParallelFileSystemAccess implements IFileSystemAccess, IFileSystemAccessExtension, IFileSystemAccessExtension2, IFileSystemAccessExtension3, IFileSystemAccessExtension4 {
 
 	Delta delta
-
```
Comment (szarnekow) [inline plugins/org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/ParallelFileSystemAccess.xtend]: Please avoid reformatting entire files in the future as it makes reviews more time consuming than necessary.

File: plugins/org.eclipse.xtext/src/org/eclipse/xtext/generator/IFileSystemAccessExtension4.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+/**
+ * This interface extends {@link IFileSystemAccess} with the capability to check files for existence.
+ *
+ * @author Karsten Thoms - Initial contribution and API
+ * @since 2.9
+ */
+public interface IFileSystemAccessExtension4 {
+
+	/**
+	 * Tests whether the file exists at the location denoted by the output configuration.
```
Comment (szarnekow) [inline plugins/org.eclipse.xtext/src/org/eclipse/xtext/generator/IFileSystemAccessExtension4.java]: Please add a note to the javadoc that directories will return false

File: tests/org.eclipse.xtext.builder.tests/src/org/eclipse/xtext/builder/builderState/EclipseResourceFileSystemAccess2Test.java [hunk 6]
```diff
@@ -28,23 +29,28 @@
...
+		fsa = injector.getInstance(EclipseResourceFileSystemAccess2.class);
 		fsa.setProject(project);
 		fsa.setOutputPath("src-gen");
 		fsa.getOutputConfigurations().get(IFileSystemAccess.DEFAULT_OUTPUT).setCreateOutputDirectory(true);
 		fsa.setMonitor(new NullProgressMonitor());
+	}
 
+	@After
+	public void tearDown() throws Exception {
+		IResourcesSetupUtil.cleanWorkspace();
```
Comment (szarnekow) [inline tests/org.eclipse.xtext.builder.tests/src/org/eclipse/xtext/builder/builderState/EclipseResourceFileSystemAccess2Test.java]: Minor: Some test classes share a common test project to speed up test execution (not in this particular test suite, though). Recently we tried to avoid a full clean of the workspace and rather remove the projects that a test class created.

**Comment by svenefftinge** [comment]
Returning false for an existing folder is very counterintuitive. We should change the semantics such that it is the same as in 'java.io.File#exists()', i.e. 

     * @return  <code>true</code> if and only if the file or directory denoted
     *          by this abstract pathname exists; <code>false</code> otherwise

That said, it might be the right time to also add 'isDirectory'.

Please squash the commits to a single one, if they all belong together. That makes reviewing much easier.

**Comment by szarnekow** [comment]
> Returning false for an existing folder is very counterintuitive. 

There is one impl of the interface that doesn't know about folders thus could not implement the contract. What's wrong with returning false for folders? IFileSystemAccess cannot deal with folders anyway as far as I can tell, e.g. there is no mkdir or similar.

**Comment by svenefftinge** [comment]
> There is one impl of the interface that doesn't know about folders thus could not implement the contract. 
Which one is that?
> What's wrong with returning false for folders?
As I said, it is against my expectations as I'm familiar with File#exists.

**Comment by szarnekow** [comment]
It's org.eclipse.xtext.generator.InMemoryFileSystemAccess

If we want to handle folders, too with the IFileSystemAccess API, we should revise it holistically, e.g. add something to access the members of a folder, mkdir, isDirectory, isFile etc. At least isDirectory would be necessary since exists() followed by read() would still throw an IOException if exists() returns true for folders.

**Comment by svenefftinge** [comment]
Yes, isDirectory should be added as well.

**Comment by szarnekow** [comment]
Alternatively we only add #isFile and don't add #exists or #isDirectory

**Comment by svenefftinge** [comment]
Yes, that would be ok for me as well.

**Comment by svenefftinge** [comment]
It really is 'isFileAndExists' which somehow tells me it should be two methods.

**Comment by szarnekow** [comment]
After an offline discussion and a look at the semantics of java.io.File.isFile, we decided to add isFile rather than exists() to the service. Karsten, do you want to update your PR?

**Comment by kthoms** [comment]
I agree that isFile does what the user wants to do, but it is not intuitive to use the method isFile if the user actually wants to prove the existence of a file. When a user searches the API for such a method he would have to dig the javadocs to find that isFile actually does what is intended, and in the user code this would need some comment that isFile is used to check for existence. I would like to avoid this.

**Comment by kthoms** [comment]
Is it now as requested?

**Comment by szarnekow** [comment]
Please remove the two files that don't belong to this PR.

**Comment by szarnekow** [comment]
Jenkins test
